components/notebook-controller/controllers/functional_test.go

```diff
@@ -0,0 +1,94 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	beta1 "github.com/kubeflow/kubeflow/components/notebook-controller/api/v1beta1"
+)
+
+var _ = Describe("Notebook controller", func() {
+
+	// Define utility constants for object names and testing timeouts/durations and intervals.
+	const (
+		Name      = "test-notebook"
+		Namespace = "default"
+		timeout   = time.Second * 10
+		interval  = time.Millisecond * 250
+	)
+
+	Context("When validating the notebook controller", func() {
+		It("Should create replicas", func() {
+			By("By creating a new Notebook")
+			ctx := context.Background()
+			notebook := &beta1.Notebook{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      Name,
+					Namespace: Namespace,
+				},
+				Spec: beta1.NotebookSpec{
+					Template: beta1.NotebookTemplateSpec{
+						Spec: v1.PodSpec{Containers: []v1.Container{{
+							Name:  "busybox",
+							Image: "busybox",
+						}}}},
+				}}
+			Expect(k8sClient.Create(ctx, notebook)).Should(Succeed())
+
+			notebookLookupKey := types.NamespacedName{Name: Name, Namespace: Namespace}
+			createdNotebook := &beta1.Notebook{}
+
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, notebookLookupKey, createdNotebook)
+				if err != nil {
+					return false
+				}
+				return true
+			}, timeout, interval).Should(BeTrue())
+			/*
+				Checking for the underlying statefulset
+			*/
+			By("By checking that the Notebook has statefulset")
+			Eventually(func() (bool, error) {
+				sts := &appsv1.StatefulSet{ObjectMeta: metav1.ObjectMeta{
+					Name:      Name,
+					Namespace: Namespace,
+				}}
+				err := k8sClient.Get(ctx, notebookLookupKey, sts)
+				if err != nil {
+					return false, err
+				}
+				return true, nil
+			}, timeout, interval).Should(BeTrue())
+			/*
+				Checking for replica count
+			*/
+			By("By checking that the Notebook is running with at least 1 replica")
+			Eventually(func() bool {
+				return notebook.Status.ReadyReplicas > 0
+			}, timeout, interval).ShouldNot(Equal(0))
```

Comment:
Is this correct? You return a bool but then check if it equal to zero (False).
So in this case, you actually check if `notebook.Status.ReadyReplicas <= 0`.
I attempted to change this to:

```go
			By("By checking that the Notebook is running with at least 1 replica")
			Eventually(func() int32 {
				// Shouldn't we get the statefulset again?
				return notebook.Status.ReadyReplicas
			}, timeout, interval).ShouldNot(BeZero())
```
but the test failed.

Which leads me to ask, in controller-runtime's test environment, is the statefulset controller running?
If it is running, what happens to the Pods it creates? I assume that they can't be scheduled, as there is no Node for them to be scheduled to. Can you dig a bit deeper on what happens with Pods in controller-runtime's test env and if they have any suggestions?

Finally, you wrapped the replica check in an `Eventually` block, but since the block doesn't get the statefulset again, the retries are meaningless.

Reply:
I agree on Pod it creates can't be scheduled without a Node. I will check on that.

Also 🤦  on the replica check in the `Eventually` block without getting the state refreshed.
It should be more like this.

``` go
By("By checking that the Notebook is running with at least 1 replica")
			Eventually(func() int32 {
				err := k8sClient.Get(ctx, notebookLookupKey, createdNotebook)
				if err != nil {
					return 0
				}
				return createdNotebook.Status.ReadyReplicas
			}, timeout, interval).ShouldNot(BeZero())
```

If the controller runtime does not have an option to create `pod` then probably this test would have to be changed. 
But I am not confident about the controller runtime not having that. Let me check and keep you posted.

Reply:
@yanniszark  I checked on the `envtest` and confirmed that the `statefulset` controller isn't running.  Only the `api server` itself is run by default. So we cannot check the `pod`. So I am refactoring my tests to check for `statefulset`